integration/build/util_test.go

```diff
@@ -0,0 +1,134 @@
+package build
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/pkg/jsonmessage"
+	"gotest.tools/assert"
+	"gotest.tools/assert/cmp"
+)
+
+var defaultBuildOptions types.ImageBuildOptions
+
+func defaultBuildOptionsWithTag(tag string) types.ImageBuildOptions {
+	buildOptions := defaultBuildOptions
+	buildOptions.Tags = []string{tag}
+	return buildOptions
+}
+
+func withBuilderVersion(buildOptions types.ImageBuildOptions) types.ImageBuildOptions {
+	if os.Getenv("DOCKER_BUILDKIT") == "1" {
+		buildOptions.Version = types.BuilderBuildKit
+	}
+	return buildOptions
+}
+
+func init() {
+	defaultBuildOptions = withBuilderVersion(types.ImageBuildOptions{
+		Remove:      true,
+		ForceRemove: true,
+	})
+}
+
+func build(buildContext io.Reader, buildOpts types.ImageBuildOptions) (*bytes.Buffer, error) {
+	apiclient := testEnv.APIClient()
+	resp, err := apiclient.ImageBuild(context.Background(), buildContext, buildOpts)
+	if err != nil {
+		return nil, err
+	}
+	out := bytes.NewBuffer(nil)
+	_, err = io.Copy(out, resp.Body)
+	resp.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+func successfulBuild(buildContext io.Reader, buildOpts types.ImageBuildOptions, msg ...string) cmp.Comparison {
+	return func() cmp.Result {
+		out, err := build(buildContext, buildOpts)
+		if err != nil {
+			return cmp.ResultFromError(err)
+		}
+		return successful(out, msg...)()
+	}
+}
+
+func successful(out *bytes.Buffer, msg ...string) cmp.Comparison {
+	return func() cmp.Result {
+		if strings.Contains(out.String(), `"errorDetail":{`) {
+			message := ""
+			if len(msg) != 0 {
+				message = " (" + strings.Join(msg, ", ") + ")"
+			}
+			return cmp.ResultFailure("build failed" + message + ", response: " + out.String())
+		}
+		return cmp.ResultSuccess
+	}
+}
+
+func buildContainerIdsFilter(buildOutput io.Reader) (filters.Args, error) {
+	const intermediateContainerPrefix = " ---> Running in "
+	filter := filters.NewArgs()
+
+	dec := json.NewDecoder(buildOutput)
+	for {
+		m := jsonmessage.JSONMessage{}
+		err := dec.Decode(&m)
+		if err == io.EOF {
+			return filter, nil
+		}
+		if err != nil {
+			return filter, err
+		}
+		if ix := strings.Index(m.Stream, intermediateContainerPrefix); ix != -1 {
+			filter.Add("id", strings.TrimSpace(m.Stream[ix+len(intermediateContainerPrefix):]))
+		}
+	}
+}
+
+func writeTarRecord(t *testing.T, w *tar.Writer, fn, contents string) {
```

Comment:
not introduced in this PR, but just noticed this should probably have `t.Helper()`?